controller/konnect/reconciler_kongplugin.go

```diff
@@ -242,31 +251,189 @@ func (r *KongPluginReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+	if err := clientWithNamespace.List(ctx, &kongServiceList); err != nil {
+		return err
+	}
+	serviceMap := lo.SliceToMap(
+		kongServiceList.Items,
+		func(obj configurationv1alpha1.KongService) (string, configurationv1alpha1.KongService) {
+			return obj.Name, obj
+		},
+	)
+
+	kongRouteList := configurationv1alpha1.KongRouteList{}
+	if err := clientWithNamespace.List(ctx, &kongRouteList); err != nil {
+		return err
 	}
+	routeMap := lo.SliceToMap(
+		kongRouteList.Items,
+		func(obj configurationv1alpha1.KongRoute) (string, configurationv1alpha1.KongRoute) {
+			return obj.Name, obj
+		},
+	)
+
+	pluginBindingsToDelete := make(map[types.NamespacedName]configurationv1alpha1.KongPluginBinding)
+	for _, pb := range kongPluginBindings {
+		// If the KongPluginBinding has a deletion timestamp, do not delete it.
+		if !pb.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		// If the KongPluginBinding is unmanaged (created not using an annotation), skip it, do not delete it.
+		if !lo.ContainsBy(pb.OwnerReferences, ownerRefIsPlugin(kongPlugin)) {
+			continue
+		}
+
+		cpRef := pb.Spec.ControlPlaneRef
+		if cpRef == nil ||
+			cpRef.KonnectNamespacedRef == nil ||
+			cpRef.Type != configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef {
+			continue
+		}
 
-	for _, b := range bindings {
-		serviceRef := b.Spec.Targets.ServiceReference
-		if serviceRef != nil &&
-			serviceRef.Group == configurationv1alpha1.GroupVersion.Group &&
-			serviceRef.Kind == "KongService" {
-			ret.byServiceName[serviceRef.Name] = append(ret.byServiceName[serviceRef.Name], b)
+		combinations, ok := groupedCombinations[types.NamespacedName{
+			// TODO: implement cross namespace references
+			Namespace: pb.Namespace,
+			Name:      cpRef.KonnectNamespacedRef.Name,
+		}]
+		if !ok {
+			pluginBindingsToDelete[client.ObjectKeyFromObject(&pb)] = pb
+			continue
 		}
 
-		routeRef := b.Spec.Targets.RouteReference
-		if routeRef != nil &&
-			routeRef.Group == configurationv1alpha1.GroupVersion.Group &&
-			routeRef.Kind == "KongRoute" {
-			ret.byRouteName[routeRef.Name] = append(ret.byRouteName[routeRef.Name], b)
+		// If the konghq.com/plugins annotation is not present, it doesn't contain
+		// the plugin in question or the object referring to the plugin has a non zero deletion timestamp,
+		// we need to delete all the managed KongPluginBindings that reference the object.
+
+		serviceRef := pb.Spec.Targets.ServiceReference
+		routeRef := pb.Spec.Targets.RouteReference
+		switch {
+
+		case serviceRef != nil && routeRef != nil:
+			combinationFound := false
+			for _, rel := range combinations {
+				if rel.Service != serviceRef.Name &&
+					rel.Route != routeRef.Name {
+					continue
+				}
+				combinationFound = true
+				break
+			}
+
+			if !combinationFound {
+				pluginBindingsToDelete[client.ObjectKeyFromObject(&pb)] = pb
+				continue
+			}
+
+			s, serviceExists := serviceMap[serviceRef.Name]
+			r, routeExists := routeMap[routeRef.Name]
```

Comment:
I believe that instead of listing all the services and routes in the namespace, we could just `Get` them one by one here and achieve the same effect without a need to convert lists to maps, etc. We could encapsulate `Get`s so it's more or less equivalent of the current code:

```go
s, serviceExists := get[*configurationv1alpha1.KongService](ctx, clientWithNamespace, types.NamespacedName{Namespace: pb.Namespace, Name: serviceRef.Name})
...
func get[T client.Object](ctx context.Context, c client.Client, key types.NamespacedName) (T, bool) {
	var obj T
	err := c.Get(ctx, key, obj)
	return obj, err == nil
}
```
WDYT?

Reply:
Something like that should work:
```go
		serviceRef := pb.Spec.Targets.ServiceReference
		routeRef := pb.Spec.Targets.RouteReference
		
		s, serviceExists := getIfRefNotNil[*configurationv1alpha1.KongService](ctx, clientWithNamespace, serviceRef)
		r, routeExists := getIfRefNotNil[*configurationv1alpha1.KongRoute](ctx, clientWithNamespace, routeRef)
```

```go
func getIfRefNotNil[T client.Object](ctx context.Context, c client.Client, ref *configurationv1alpha1.TargetRefWithGroupKind) (T, bool) {
	var obj T
	if ref == nil {
		return obj, false
	}
	err := c.Get(ctx, client.ObjectKey{Name: ref.Name} , obj)
	return obj, err == nil
}
```

Reply:
Makes sense 👍 I was wondering if listing and then mapping vs just getting would be better (given that we use a cached client) but without proper performance tests that's just speculation I guess.

Implemented the above in 2321541450fea901a9d75f28596cfdcf99ade0d9 (Needed to adjust the type params a bit)